### PR TITLE
feat: create2 for proxy 

### PR DESCRIPTION
**Pull request by xmxanuel**
- use create2 for proxy
- removal of proxies mapping

- removed the proxy constructor parameter and used a init method

 I first implemented it with constructor parameter (they are just attached to the bytecode for the create) but it would change the hash of the bytecode and we would have to recalculate it every time for the proxies method. (expensive)

### CODE REVIEW

**Review by charleenfei — APPROVED** [review]

**Review by ilinzweilin — APPROVED** [review]

File: src/proxy.sol [hunk 1]
```diff
@@ -84,10 +91,23 @@ contract Proxy is TitleOwned {
...
+        // create2 address calculation
+        // keccak256(0xff ++ deployingAddr ++ salt ++ keccak256(bytecode))[12:]
+        bytes32 _data = keccak256(
+            abi.encodePacked(bytes1(0xff), address(this), keccak256(abi.encodePacked(accessToken)), proxyCodeHash)
+        );
+        return address(bytes20(_data << 96));
+    }
 
     constructor() Title("Tinlake Actions Access Token", "TAAT") public {
+        proxyCodeHash = keccak256(abi.encodePacked(proxyCode));
```
Comment (lucasvo) [inline src/proxy.sol:110]: if the bytecode is hardcoded, I think we should hardcode this as well.

File: src/proxy.sol [hunk 2]
```diff
@@ -98,10 +118,22 @@ contract ProxyRegistry is Title {
...
-        emit Created(msg.sender, owner, proxy, token);
+        uint accessToken = _issue(owner);
+        bytes32 salt = keccak256(abi.encodePacked(accessToken));
+        proxy = deploy(accessToken, salt);
+        emit Created(msg.sender, owner, proxy, accessToken);
+    }
+
+
+    /// uses the create2 opcode to deploy a proxy contract
+    function deploy(uint accessToken, bytes32 salt) internal returns (address payable addr) {
```
Comment (lucasvo) [inline src/proxy.sol:129]: If deploy is only called from the build method, why not merge the two functions?

File: src/proxy.sol [hunk 3]
```diff
@@ -30,9 +30,16 @@ contract Proxy is TitleOwned {
     uint public         accessToken;
     RegistryLike public registry;
 
-    constructor(uint accessToken_) TitleOwned(msg.sender) public {
+    bool initialized;
```
Comment (lucasvo) [inline src/proxy.sol:33]: This can be removed and replaced by a non zero check for accessToken.

File: src/proxy.sol [hunk 4]
```diff
@@ -30,9 +30,16 @@ contract Proxy is TitleOwned {
...
+
+    constructor() TitleOwned(msg.sender) public {
         registry = RegistryLike(msg.sender);
+    }
+
+    function init(uint accessToken_) public {
+        require(initialized == false);
```
Comment (lucasvo) [inline src/proxy.sol:40]: Replace with:
         require(accessToken == 0);

File: src/proxy.sol (hunk 1 above)
Comment (lucasvo) [reply src/proxy.sol:110]: On second thought, why don't we pass the contract bytecode into the constructor?

Comment (xmxanuel) [reply src/proxy.sol:110]: you can ensure it is the correct hash if you do it in the constructor

**Pull request merged**